### DISCUSSION TITLE
skywalkingagent SamplingService

### DISCUSSION THREAD

**Discussion by BritneyStars** [apache/skywalking#10549] [Q&A] [answered]
In thread monitor jstack, these are lots of skywalking threads such as:

Full thread dump Open]DK 64-Bit Server VM (25.322-b06 mixed mode):
"SkywalkingAgent-1082-SamplingService-0" #15601 daemon prio=5 os_prio=0
tid=0x00007f25e874d000 nid=0xc4d0 waiting o condition [0x00007f23cee72000]
java.lang.Thread.State:TIMED_WAITING(parking)
at sun.misc. Unsafe.park(Native Method)
- parking to wait for <0x000000070bd46ac0> (a
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
atjava.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
at
java.util.concurrent.locks.AbstractQueuedSynchronizersConditionObject.awaitNanos(Ab
stractQueuedSynchronizer.java:2078)
at
java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(Scheduled
ThreadPoolExecutor java:1093)
au
java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(Scheduled
ThreadPoolExecutor.java:809)
at
java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1074)
at
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
at
java.util.concurrent.ThreadPoolExecutorsWorker.run(ThreadPoolExecutor.java:624)
atjava.lang.Thread.run(Thread.java:750)
Locked ownable synchronizers:

It seems that when I change the sample rate of agentConfiguration oap server , There will create new thread. Many threads is in wailting state.why?

**Comment by wu-sheng** [comment]
There is on and off switch here https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/sampling/SamplingService.java#L116

You could debug to see whether there is a thread leak?

**Reply by BritneyStars** [reply]
I debug in local enviroment .but found lots of errors about java.lang.NoSuchMethodError:org.apache.skywaling.apm.network.language.agent.v3.ConfigurationDiscoveryServiceGrpc.newBlockingStub

**Reply by wu-sheng** [reply]
This means your backend is not providing this service, maybe due to older versions?

**Reply by BritneyStars** [reply]
I use the V9.3.0 backend version,agent is v8.13.

**Reply by BritneyStars** [reply]
the whole error is :
java. Lang. NosuchMethodError: org. apache. skywalking. apm. network. language. agent. v3 .ConfigurationDiscoveryServiceGrpc.newBlockingstub (Lorg/apache/skywalking/apm/dependencies/io/grpc/Channel;) Lorg/apache/skywalking/apm/network/language/agent/v3/ConfigurationDiscoveryServiceGrpe$ConfigurationDiscoveryServiceBlockingStub;   why  package name is start with Lorg? it should be org? whether windows system cause this?

**Reply by wu-sheng** [reply]
L usually means array at bytecode level.

**Reply by BritneyStars** [reply]
I found the error, because  my project also use the grpc function and the grpc version is different from the agent. I debug in local  and temporarily do not find that Samplingservices there is a thread leak 。But in my production enviroment, there are more than 1000 threads -SamplingService-threads. I will observe times

**Comment by hyhyf** [comment]
I found the same error, but for different reasons. In my production enviroment, when agent.sample_n_per_3_secs keeps switching back and forth between -1 and other numbers, SampleService will always create new threads, Over time, the thread leaked.
How about code like this?
`

    private ScheduledExecutorService service;
    @Override
    public void boot() {
        service = Executors.newSingleThreadScheduledExecutor(
                new DefaultNamedThreadFactory("SamplingService"));
        samplingRateWatcher = new SamplingRateWatcher("agent.sample_n_per_3_secs", this);
        ServiceManager.INSTANCE.findService(ConfigurationDiscoveryService.class)
                               .registerAgentConfigChangeWatcher(samplingRateWatcher);

        handleSamplingRateChanged();
    }

    public void handleSamplingRateChanged() {
        if (samplingRateWatcher.getSamplingRate() > 0) {
            if (!on) {
                on = true;
                this.resetSamplingFactor();
                scheduledFuture = service.scheduleAtFixedRate(new RunnableWithExceptionProtection(
                        this::resetSamplingFactor, t -> LOGGER.error("unexpected exception.", t)), 0, 3, TimeUnit.SECONDS);
                LOGGER.debug(
                        "Agent sampling mechanism started. Sample {} traces in 3 seconds.",
                        samplingRateWatcher.getSamplingRate()
                );
            }
        } else {
            if (on) {
                if (scheduledFuture != null) {
                    scheduledFuture.cancel(true);
                }
                on = false;
            }
        }
    }`

**Reply by hyhyf** [reply]
@wu-sheng Please confirm, thx.

**Reply by wu-sheng** [reply]
Please submit a pull request and explain why it is an issue, and how you fix this.
We should discuss at there.

**Comment by wu-sheng** [comment, marked as answer]
Fixed through https://github.com/apache/skywalking-java/pull/487
